src/Futhark/CLI/Eval.hs

```diff
@@ -0,0 +1,155 @@
+module Futhark.CLI.Eval (main) where
+  
+import Control.Exception
+import Control.Monad
+import Control.Monad.Except
+import Control.Monad.Free.Church
+
+import Data.Maybe
+import Data.Text qualified as T
+
+import Futhark.Compiler
+import Futhark.MonadFreshNames
+-- import Futhark.Util (toPOSIX)
+import Futhark.Util.Options
+
+import Language.Futhark
+import Language.Futhark.Interpreter qualified as I
+import Language.Futhark.Parser
+import Language.Futhark.Semantic qualified as T
+import Language.Futhark.TypeChecker qualified as T
+
+import System.FilePath
+
+import Futhark.Pipeline
+
+import System.IO
+import Prelude
+-- import Language.Futhark.Interpreter (interpretExp, ExtOp)
+
+
+import Futhark.Util.Pretty
+import Data.List (foldl')
+
+main :: String -> [String] -> IO ()
+-- main _ _ = putStrLn "Hello World"
+main = mainWithOptions interpreterConfig options "options... <exprs...>" run
+  where
+    run [] _ = Nothing
+    run exprs config = Just $ runExprs exprs config
+
+runExprs :: [String] -> InterpreterConfig -> IO ()
+runExprs exprs cfg = do
+  -- putStrLn "Running expressions"
+  let InterpreterConfig _ file = cfg
+  -- putStrLn "File has come up."
+  maybe_new_state <- newFutharkiState cfg file
+  -- putStrLn "New state made"
+  (src, env, ctx) <- case maybe_new_state of
+    Left _ -> error "Unable to load file."
+    Right s -> pure s
+  -- putStrLn "Starting on the expressions"
+  foldl' (\_ b -> runExpr src env ctx b) (putStr "") exprs
+  putStrLn ""
+-- runExprs _ _ = putStrLn "Running Exprs"
+
+
+
+-- Use parseExp, checkExp, then interpretExp.
+runExpr :: VNameSource -> T.Env -> I.Ctx -> String -> IO ()
+runExpr src env ctx str = do
+  uexp <- case parseExp "" (T.pack str) of 
+    Left _ -> error "Syntax Error"
+    Right e -> pure e
+  let (_, expdat) = T.checkExp [] src env uexp
+  fexp <- case expdat of
+    Left _ -> error "Type Error"
+    Right (_,e) -> pure e
+  let ext = I.interpretExp ctx fexp
+  pval <- runInterpreter' ext
+  val <- case pval of
+    Left _ -> error ""
```

Comment:
Same as above; don't bind before pattern-matching and don't use `error`.

Reply:
I am getting an error when I bind during pattern matching:
```
  val <- case runInterpreter' ext of
    Left _ -> error ""
    Right (Left _) ->  error ""
    Right (Right x) -> pure x
```
```
• No instance for (MonadIO (Either a0))
    arising from a use of ‘runInterpreter'’
• In the expression: runInterpreter' ext
  In a stmt of a 'do' block:
    val <- case runInterpreter' ext of
             Left _ -> error ""
             Right (Left _) -> error ""
             Right (Right x) -> pure x
  In the expression:
    do uexp <- case parseExp "" (T.pack str) of
                 Left _ -> error $ "Syntax Error:" <> str
                 Right e -> pure e
       fexp <- case T.checkExp [] src env uexp of
                 (_, Left _) -> error $ "Type Error:" <> str
                 (_, Right (_, e)) -> pure e
       let ext = I.interpretExp ctx fexp
       val <- case runInterpreter' ext of
                Left _ -> error ""
                Right (Left _) -> error ""
                Right (Right x) -> pure x
       ....
```

Reply:
I think he means that you shouldn't bind the result of the `case`. For example:
```
case pval of
    Left err -> do
      putDoc $ I.prettyInterpreterError err
      exitWith $ ExitFailure 0
    Right x -> putDoc $ I.prettyValue x <> hardline
```
and similar for the other cases. [Here's an example.](https://github.com/diku-dk/futhark/blob/9780afb82718ddd3c394e902f28e8a73e953044e/src/Futhark/CLI/REPL.hs#L252-L274)

Reply:
i think this one is realted to binding *before* pattern matching, but i will avoid assigning the `case` expr as well.